src/classes/CON_AddToCampaign.cls

```diff
@@ -0,0 +1,209 @@
+                MAPcampaignMemberStatusesCompare.put(Campaigns[i], new list<string>{Statuses[i]});
+            }
+        }
+        
+        //iterate through both maps comparing the values against each other and adding new campaign member values where necessary
+        for(Id cIdCompare : MAPcampaignMemberStatusesCompare.keyset()) {
+            
+            integer sortOrderMax = 3;
+            
+            for(string cmsCompare : MAPcampaignMemberStatusesCompare.get(cIdCompare)) {
+                
+                boolean match = false;
+                
+                if(MAPcampaignMemberStatuses.containsKey(cIdCompare)) {
+                    for(CampaignMemberStatus cms : MAPcampaignMemberStatuses.get(cIdCompare)) {
+                        
+                        if(sortOrderMax < cms.SortOrder)
+                            sortOrderMax = cms.SortOrder;
+                            
+                        if(cmsCompare == cms.Label) {
+                            match = true;
+                            break;
+                        }
+                    }
+                }
+                if(!match && !isSpecialStatus(cmsCompare)) {
+                    sortOrderMax ++;
+                    listStatusForInsert.add(new CampaignMemberStatus(campaignId = cIdCompare,
+                                                                Label = cmsCompare,
+                                                                HasResponded = false,
+                                                                SortOrder = sortOrderMax));
+                }
+            }
+        }
+
+        insert listStatusForInsert;
+    }
+        
+    /*******************************************************************************************************
+    * @description With all the associated processing complete, this method creates and updates campaign members
+    * @param Contacts the list of Contact IDs.
+    * @param Campaigns the list of Campaign IDs.
+    * @param Statuses the list of Status strings.
+    * @return void This method adds campaign members to dmlWrapper and returns nothing.
+    ********************************************************************************************************/
+    private static void getCampaignMembers(list<id> Contacts, list<id> Campaigns, list<String> Statuses) {
+        
+        //iterate the list of Contacts passed in and check for matches in  Campaign Members and if found update Campaign Member Status
+        for(integer i = 0; i < Contacts.size(); i++) {
+            string comboKey = Contacts[i] + ':' + Campaigns[i];
+
+            //if we have an existing campaign member, update the status
+            if(MAPCampaignMembers.containsKey(comboKey)) {
+                CampaignMember cmForUpdate = MAPCampaignMembers.get(comboKey);
+                cmForUpdate.status = getStatus(Campaigns[i], Statuses[i]);
+                dmlWrapper.objectsToUpdate.add(cmForUpdate);
```

Comment:
i'm not a fan of using the dmlWrapper static directly.  how can this routine ensure it is instantiated?  wouldn't it be cleaner to have this routine take a dmlWrapper argument?